libr/bin/dwarf.c

```diff
@@ -763,71 +829,52 @@ R_API int r_bin_dwarf_parse_line_raw2(const RBin *a, const ut8 *obuf,
 	return true;
 }
 
-#define READ_BUF(x,y) if (idx+sizeof(y)>=len) { return false;} \
-	(x)=*(y*)buf; idx+=sizeof(y);buf+=sizeof(y)
-
 R_API int r_bin_dwarf_parse_aranges_raw(const ut8 *obuf, int len, FILE *f) {
-	ut32 length, offset;
-	ut16 version;
-	ut32 debug_info_offset;
-	ut8 address_size, segment_size;
 	const ut8 *buf = obuf;
 	int idx = 0;
 
-	if (!buf || len< 4) {
+	if (!buf || len < 4) {
 		return false;
 	}
 
-	READ_BUF (length, ut32);
-	if (f) {
-		printf("parse_aranges\n");
-		printf("length 0x%x\n", length);
-	}
-
-	if (idx + 12 >= len) {
-		return false;
-	}
+	// FIXME endian
+	data_xtor ex;
+	data_xtor_init (&ex, obuf, len, false);
+	st64 off = 0;
 
-	READ_BUF (version, ut16);
+	ut32 length = data_xtor_u32 (&ex, &off);
+	ut16 version = data_xtor_u16 (&ex, &off);
+	ut32 cu_offset = data_xtor_u32 (&ex, &off);
+	ut8 addr_size = data_xtor_u8 (&ex, &off);
+	ut8 seg_size = data_xtor_u8 (&ex, &off);
 	if (f) {
```

Comment:
the helper function is not returning from the parent function in case of failing, just returns 0; which is wrong. this can be only done with a macro or by doing some spaguetti, anyway this can be also simplfied by using sdb_fmt apis or the r_buf_fcpy apis, because this data_xtor struct is doing the same as RBuffer